flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecTableSourceScan.java

```diff
@@ -85,10 +85,16 @@ public DynamicTableSourceSpec getTableSourceSpec() {
         } else if (provider instanceof SourceProvider) {
             Source<RowData, ?, ?> source = ((SourceProvider) provider).createSource();
             // TODO: Push down watermark strategy to source scan
-            return env.fromSource(source, WatermarkStrategy.noWatermarks(), operatorName)
-                    .getTransformation();
+            Transformation<RowData> transformation =
+                    env.fromSource(source, WatermarkStrategy.noWatermarks(), operatorName)
+                            .getTransformation();
+            transformation.setOutputType(outputTypeInfo);
+            return transformation;
```

Comment:
We can use `env.fromSource(source, WatermarkStrategy.noWatermarks(), operatorName, outputTypeInfo)` for this.